### PR TITLE
Type hints for class based views (and other tools)

### LINKED ISSUE #2055: Type-check class based views from #1964
In #1964, we made a mistake with the instance variables and return types inside [`UserMergeSelectionView::get_success_url`](https://github.com/e-valuation/EvaP/pull/1964/files#diff-94e55a2578328edb5cc0dd3b57b2ceae72f5e824fff0742ae770d59451f046d7R2292). Mypy didn't complain about these mistakes. I would have expected that the function signature of `get_success_url` from the base class is applied, but [mypy doesn't do this](https://mypy.readthedocs.io/en/stable/class_basics.html#overriding-statically-typed-methods).

To make mypy detect such errors, I think we want to type-annotate the class based views introduced in #1964 and possibly also mark them with the `@override` decorator.

### PR DESCRIPTION
**Pull request by richardebeling**
Fixes #2055

### CODE REVIEW

File: evap/evaluation/tools.py [hunk 1]
```diff
@@ -9,10 +9,12 @@
 from django.conf import settings
 from django.core.exceptions import SuspiciousOperation, ValidationError
 from django.db.models import Model
-from django.http import HttpResponse
+from django.forms.formsets import BaseFormSet
+from django.http import HttpRequest, HttpResponse
 from django.shortcuts import get_object_or_404
 from django.utils.translation import get_language
 from django.views.generic import FormView
+from typing_extensions import Protocol
```
Comment (niklasmohrin) [inline evap/evaluation/tools.py:17]: `Protocol` exists in `typing` since 3.8 :)

File: evap/evaluation/tools.py [hunk 2]
```diff
@@ -58,52 +60,52 @@ def is_prefetched(instance, attribute_name: str):
...
 
-def is_external_email(email):
+def is_external_email(email: str) -> bool:
     return not any(email.endswith("@" + domain) for domain in settings.INSTITUTION_EMAIL_DOMAINS)
 
 
-def sort_formset(request, formset):
+def sort_formset(request: HttpRequest, formset: BaseFormSet) -> None:
     if request.POST:  # if not, there will be no cleaned_data and the models should already be sorted anyways
         formset.is_valid()  # make sure all forms have cleaned_data
```
Comment (niklasmohrin) [inline evap/evaluation/tools.py:90]: (not yours), but?
```suggestion
        assert formset.is_valid()  # make sure all forms have cleaned_data
```

File: evap/evaluation/tools.py [hunk 3]
```diff
@@ -115,7 +117,7 @@ def translate(**kwargs):
     return property(lambda self: getattr(self, kwargs[get_language() or "en"]))
 
 
-def clean_email(email):
+def clean_email(email: str) -> str:
```
Comment (niklasmohrin) [inline evap/evaluation/tools.py:120]: This is actually `str | None`, see for example `UserForm.clean_email`. Not sure if we can model a `None` iff `None` relationship with generics or so, would be cool

**Review by niklasmohrin — CHANGES_REQUESTED** [review]
Everything else looks good to me

File: evap/evaluation/tools.py (hunk 2 above)
Comment (richardebeling) [reply evap/evaluation/tools.py:90]: One test fails then. The method seems to be intended to also be callable with invalid formsets (via the default value in the line below).

**Review by niklasmohrin — APPROVED** [review]

**Pull request merged**